compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+
+    result[pragmasPos] =
+      case n[pragmasPos].kind
+      of nkEmpty:  empty(n[pragmasPos].info)
+      of nkPragma: parsePragma(c, n[pragmasPos])
+      else:        invalidAst(c, n[pragmasPos])
+
+    # ignore the contents of the 'misc' slot
+    result[miscPos] = empty(n.info)
+    result[bodyPos] = stmt(c, n[bodyPos])
+    if n.len > resultPos:
+      result[resultPos] = empty(n.info)
+
+  of nkOfBranch, nkElifBranch, nkExceptBranch, nkElse, nkElifExpr, nkElseExpr:
+    invalid()
+  of nkAsmStmt:
+    checkSonsLen(n, 2)
+    result = newTreeI(nkAsmStmt, n.info)
+    result[0] = emptyOr(c, n, parsePragma)
+    result[1] = parseStrLit(c, n[1])
+  of nkPragma:
+    checkMinSonsLen(n, 1)
+    for i in 0..<n.len:
+      result[i] = parseColonEqExpr(c, n[i])
+  of nkPragmaBlock:
+    unreachable("missing")
+  of nkForStmt:
+    checkMinSonsLen(n, 3)
+    result = prepareFrom(n)
+    # note: this allows for code that the parser doesn't
+    for i, it in sliceIt(n, 0..^3):
+      result[i] =
+        case it.kind
+        of nkVarTuple:   parseVarTuple(c, it)
+        of nkPostfix:    parseIdentVis(c, it)
+        of nkPragmaExpr: parsePragmaExpr(c, it)
+        else:            parseIdent(c, it)
+
+    result[^2] = expr(c, n[^2])
+    result[^1] = stmt(c, n[^1])
+  of nkWhileStmt:
+    checkSonsLen(n, 2)
+    result = newTreeI(nkWhileStmt, n.info): [expr(c, n[0]), stmt(c, n[1])]
+  of nkCaseStmt:
+    checkMinSonsLen(n, 2)
+    result = newNodeI(nkCaseStmt, n.info, n.len)
+    result[0] = expr(c, n[0])
+    for i, it in sliceIt(n, 1..^1):
+      result[i] =
+        case it.kind
+        of nkOfBranch: parseBranch(c, it, true)
+        of nkElifBranch, nkElifExpr: parseBranch(c, it, true)
+        of nkElse, nkElseExpr:
+          guardLen(it, 1):
+            newTreeI(it.kind, it.info): expr(c, it[0])
+        else: invalidAst(c, it)
+  of nkTypeSection:
+    unreachable("missing")
+  of nkVarSection, nkLetSection:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] = parseVariable(c, n[i])
+  of nkConstSection:
+    unreachable("missing")
+  of nkConstDef, nkTypeDef:
+    # context dependent
+    invalid()
+  of nkDefer:
+    unreachable("missing")
+  of nkYieldStmt, nkRaiseStmt, nkReturnStmt, nkBreakStmt, nkContinueStmt, nkDiscardStmt:
+    checkSonsLen(n, 1)
+    result = prepareFrom(n)
+    result[0] =
+      if n[0].kind == nkEmpty: empty(n[0].info)
+      else:                    expr(c, n[0])
+
+  of nkFinally:
+    invalid()
+  of nkBlockStmt:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = stmt(c, n[1])
+  of nkStaticStmt:
+    checkSonsLen(n, 1)
+    result = stmt(c, n[0])
+  of nkStmtList:
+    result = stmts(c, n)
+  of nkImportStmt, nkImportExceptStmt, nkExportStmt, nkExportExceptStmt, nkFromStmt, nkIncludeStmt:
+    # QUESTION: should we perform any additional validation here or leave that
```

Comment:
most of the validation should remain in sem, doing some basics here for sanity is fine, but I wouldn't "hurt" ourselves to do it here.